### ISSUE TITLE
What is the next plan of FP8 support in PyTorch?

### ISSUE THREAD

**Issue by yanbing-j** [pytorch/pytorch#117490]
### 🚀 The feature, motivation and pitch

Now PyTorch only supports FP8 data type conversion without scaling. The accuracy is not that good.

What is the plan of FP8 support in PyTorch? Will FP8 DelayedScaling from TransformerEngine be taken into account? Thanks!

cc @svekars @brycebortree @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel @albanD @kadeng

**Comment by malfet** [comment]
See https://github.com/pytorch-labs/float8_experimental

**Comment by vkuzo** [comment]
Hi @yanbing-j , please see https://dev-discuss.pytorch.org/t/float8-in-pytorch-1-x/1815 for context on float8 training in native PyTorch

**Comment by yanbing-j** [comment]
Thank @malfet @vkuzo for the detailed explaination of fp8_experimental. Will fp8_experimental be merged into PyTorch main branch in the future?

I also have a question of the fp8 autograd. The document mentions that, to get around the restriction of "x.dtype to equal x.grad.dtype", we use Float8Tensor. In my view, the input of FP8Linear could be bfloat16, for FP8Linear operator, its grad dtype should be bfloat16, and output dtype should also be bfloat16, right? And x.dtype becomes fp8 dtype, because of the cast operation in FP8Linear forward. This supposes not to influence autograd, since TransformerEngine uses `class _Linear(torch.autograd.Function)`. So is `Float8Tensor` necessary in PyTorch implementation?

Thanks!

**Comment by vkuzo** [comment]
> Will fp8_experimental be merged into PyTorch main branch in the future?

We will find a place for it once the codebase is more stable.  Parts of it might be eventually be in core, and parts of it might be in a standalone repository.

> This supposes not to influence autograd, since TransformerEngine uses class _Linear(torch.autograd.Function). So is Float8Tensor necessary in PyTorch implementation?

Using Float8Tensor is necessary for mixed dtypes for fw/bw when outside of `torch.autograd.Function`.  One can do mixed dtypes inside `torch.autograd.Function`, but that is hard to compose cleanly with other things as then one is forced to have an all-encompassing `torch.autograd.Function` with a high number of input arguments.

**Comment by yanbing-j** [comment]
Understand. Thank you @vkuzo .
